src/libstore/gc.cc

```diff
@@ -778,107 +788,108 @@ void LocalStore::collectGarbage(const GCOptions & options, GCResults & results)
 
     /* Either delete all garbage paths, or just the specified
        paths (for gcDeleteSpecific). */
-    if (options.action == GCOptions::gcDeleteSpecific) {
-
-        for (auto & i : options.pathsToDelete) {
-            deleteReferrersClosure(i);
-            if (!dead.count(i))
-                throw Error(
-                    "Cannot delete path '%1%' since it is still alive. "
-                    "To find out why, use: "
-                    "nix-store --query --roots",
-                    printStorePath(i));
-        }
-
-    } else if (options.maxFreed > 0) {
-
-        if (shouldDelete)
-            printInfo("deleting garbage...");
-        else
-            printInfo("determining live/dead paths...");
-
-        try {
-            AutoCloseDir dir(opendir(realStoreDir.get().c_str()));
-            if (!dir) throw SysError("opening directory '%1%'", realStoreDir);
-
-            /* Read the store and delete all paths that are invalid or
-               unreachable. We don't use readDirectory() here so that
-               GCing can start faster. */
-            auto linksName = baseNameOf(linksDir);
-            Paths entries;
-            struct dirent * dirent;
-            while (errno = 0, dirent = readdir(dir.get())) {
-                checkInterrupt();
-                std::string name = dirent->d_name;
-                if (name == "." || name == ".." || name == linksName) continue;
+    std::visit(overloaded{
+        [&](GCDelete del){
+            if (del.pathsToDelete.has_value()) {
+                for (auto & i : del.pathsToDelete->paths) {
+                    deleteReferrersClosure(i);
+                    if (!del.pathsToDelete->recursive && !dead.count(i))
+                        throw Error(
+                            "Cannot delete path '%1%' since it is still alive. "
+                            "To find out why, use: "
+                            "nix-store --query --roots",
+                            printStorePath(i));
+                }
+            } if (options.maxFreed > 0) {
```

Comment:
I don't think this does what it should:

- Even if `recursive` is set, it won't recurse into the (references) closure of the path. All it will do is not fail if the path can't be deleted
- It seems (both from the code and from testing) that it will always trigger a full GC, even when a partial one (closure or single path) is requested

Reply:
> It seems (both from the code and from testing) that it will always trigger a full GC, even when a partial one (closure or single path) is requested

Sorry, that was an off-by-one bug, fixed now.

> Even if recursive is set, it won't recurse into the (references) closure of the path. All it will do is not fail if the path can't be deleted

Whoops! I lost the bit of code I wrote to GC recursively...